agent/internal/fluent.go

```diff
@@ -0,0 +1,450 @@
+			Content:  []byte(parserConfig),
+		},
+	)
+
+	baseConfig := fmt.Sprintf(`
+[SERVICE]
+  # Flush every .05 seconds to reduce latency for users.
+  Flush .05
+  Parsers_File %s
+
+[INPUT]
+  Name forward
+`, parserConfigPath)
+
+	filterConfig := fmt.Sprintf(`
+# Attempt to parse the rank ID and log level out of output lines.
+[FILTER]
+  Name parser
+  Match *
+  Key_Name log
+  Parser rank_id
+  Reserve_Data true
+
+[FILTER]
+  Name parser
+  Match *
+  Key_Name log
+  Parser log_level
+  Reserve_Data true
+
+# Move around fields to create the desired shape of object.
+[FILTER]
+  Name modify
+  Match *
+  # Delete Docker's container information, which we don't want.
+  Remove container_id
+  Remove container_name
+  # Rename environment variables to normal names.
+  Rename DET_TRIAL_ID trial_id
+  Rename DET_CONTAINER_ID container_id
+
+  Add agent_id %s
+  Rename source stdtype
+
+# Apply the Lua code for miscellaneous field tweaking.
+[FILTER]
+  Name lua
+  Match *
+  Script %s
+  Call run
+`, opts.AgentID, luaPath)
+
+	outputConfig := fmt.Sprintf(`
+# TMP: Output to stdout as well.
+[OUTPUT]
+  Name stdout
+  Match *
+
+[OUTPUT]
+  Name http
+  Match *
+  Host %s
+  Port %d
+  URI /trial_logs
+  Header_tag X-Fluent-Tag
+  Format json
+  Json_date_key timestamp
+  Json_date_format iso8601
+`, opts.MasterHost, opts.MasterPort)
+
+	if opts.Security.TLS.Enabled {
+		outputConfig += "  tls On\n"
+
+		if opts.Security.TLS.SkipVerify {
+			outputConfig += "  tls.verify Off\n"
+		}
+		if opts.Security.TLS.MasterCert != "" {
+			outputConfig += "  tls.ca_file " + masterCertPath + "\n"
+
+			certBytes, cErr := ioutil.ReadFile(opts.Security.TLS.MasterCert)
+			if cErr != nil {
+				return nil, nil, cErr
+			}
+
+			files = append(files,
+				archive.Item{
+					Path:     masterCertPath,
+					Type:     tar.TypeReg,
+					FileMode: 0444,
+					Content:  certBytes,
+				},
+			)
+		}
+	}
+
+	files = append(files,
+		archive.Item{
+			Path:     configPath,
+			Type:     tar.TypeReg,
+			FileMode: 0444,
+			Content:  []byte(baseConfig + filterConfig + outputConfig),
+		})
+
+	args := []string{"/fluent-bit/bin/fluent-bit", "-c", configPath}
+
+	return args, files, nil
+}
+
+func killContainer(ctx *actor.Context, c *client.Client, containerID string) error {
+	if err := c.ContainerKill(context.Background(), containerID, "KILL"); err != nil {
+		return errors.Wrap(err, "failed to kill container")
+	}
+	return nil
+}
+
+func killContainerByName(ctx *actor.Context, docker *client.Client, name string) error {
+	containers, err := docker.ContainerList(context.Background(), types.ContainerListOptions{
+		Filters: filters.NewArgs(
+			filters.Arg("name", name),
+		),
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to list containers by name")
+	}
+	for _, cont := range containers {
+		ctx.Log().WithFields(logrus.Fields{"name": name, "id": cont.ID}).Infof("killing Docker container")
+		if err = killContainer(ctx, docker, cont.ID); err != nil {
```

Comment:
blocking: I think this is a race condition. Peeked at the docker daemon code and kill just sends a signal and returns but doesn't wait for anything to happen. In ~10 runs once I did actually see where we tried to create the container but it hadn't fulled stopped and been removed yet.
```
determined-agent   | ERRO[2020-10-28T10:07:15-04:00] error while actor was running                 error="failed to start Fluent daemon: Error response from daemon: Conflict. The container name \"/determined-fluent\" is already in use by container \"64ffc0be5858bef1612e40ef11f05978946d8292519179ec742ae1274272d97f\". You have to remove (or rename) that container to be able to reuse that name." id=agent system=Bradleys-MBP.attlocal.net type=agent
```
I think we need to throw a `c.ContainerWait()` behind all of these.

Reply:
Sorry I'm leaving these all 1 by 1. As I'm writing the elastic stuff, I'm getting a lot more familiar with the code than I was able to get from review.